jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -191,42 +191,54 @@ public int getMaxCapacity()
     public RetainableByteBuffer acquire(int size, boolean direct)
     {
         RetainedBucket bucket = bucketFor(size, direct);
+
+        // No bucket, return non-pooled.
         if (bucket == null)
-            return newRetainableByteBuffer(size, direct, this::removed);
+            return newRetainableByteBuffer(size, direct, null);
+
+        // Try to acquire a pooled entry.
         Pool.Entry<RetainableByteBuffer> entry = bucket.getPool().acquire();
+        if (entry != null)
+        {
+            RetainableByteBuffer buffer = entry.getPooled();
+            ((Buffer)buffer).acquire();
+            return buffer;
+        }
+
+        // No pooled entry available, try to reserve.
+        Pool.Entry<RetainableByteBuffer> reservedEntry = bucket.getPool().reserve();
+        // Cannot reserve, return non-pooled.
+        if (reservedEntry == null)
+            return newRetainableByteBuffer(size, direct, null);
 
-        RetainableByteBuffer buffer;
-        if (entry == null)
+        // Allocate the buffer that will be pooled.
+        RetainableByteBuffer buffer = newRetainableByteBuffer(bucket.getCapacity(), direct, retainedBuffer ->
         {
-            Pool.Entry<RetainableByteBuffer> reservedEntry = bucket.getPool().reserve();
-            if (reservedEntry != null)
-            {
-                buffer = newRetainableByteBuffer(bucket._capacity, direct, retainedBuffer ->
-                {
-                    BufferUtil.reset(retainedBuffer.getByteBuffer());
-                    if (!reservedEntry.release())
-                        reservedEntry.remove();
-                });
+            BufferUtil.reset(retainedBuffer.getByteBuffer());
+            release(bucket, reservedEntry, direct);
+        });
 
-                // A reserved entry may become old and be evicted before it is enabled.
-                if (reservedEntry.enable(buffer, true))
-                {
-                    if (direct)
-                        _currentDirectMemory.addAndGet(buffer.capacity());
-                    else
-                        _currentHeapMemory.addAndGet(buffer.capacity());
-                    releaseExcessMemory(direct);
-                    return buffer;
-                }
-            }
-            return newRetainableByteBuffer(size, direct, this::removed);
-        }
+        reservedEntry.enable(buffer, true);
 
-        buffer = entry.getPooled();
-        ((Buffer)buffer).acquire();
         return buffer;
     }
 
+    private void release(RetainedBucket bucket, Pool.Entry<RetainableByteBuffer> entry, boolean direct)
+    {
+        long maxMemory = direct ? _maxDirectMemory : _maxHeapMemory;
+        // Account also for the entry that is about to be released.
+        long memory = getMemory(direct) + bucket.getCapacity();
+        if (maxMemory > 0 && memory > maxMemory)
+        {
+            bucket.evict(entry);
```

Comment:
@sbordet If you move the reservation logic here, you only need to check the size if you have to reserve, as that is the only time the pool grows.

I also prefer to evict an idle buffer which is more likely to be under-utilized, rather than the buffer just released, with has just been used.

See #11382